target/linux/ath79/image/nand.mk

```diff
@@ -110,6 +111,19 @@ define Device/glinet_gl-ar750s-nor
 endef
 TARGET_DEVICES += glinet_gl-ar750s-nor
 
+define Device/mikrotik_routerboard-922uags-5hpacd
+  $(Device/mikrotik)
+  SOC := qca9558
+  DEVICE_MODEL := RouterBOARD 922UAGS-5HPacD
+  BOARD_NAME := routerboard
```

Comment:
Don't think board name is still required?

Reply:
It depends. I got the "platform_do_upgrade_mikrotik_rb" as is from ar71xx:
```
platform_do_upgrade_mikrotik_rb() {
	CI_KERNPART=none
	local fw_mtd=$(find_mtd_part kernel)
	fw_mtd="${fw_mtd/block/}"
	[ -n "$fw_mtd" ] || return
	mtd erase kernel
	tar xf "$1" sysupgrade-routerboard/kernel -O | nandwrite -o "$fw_mtd" -

	nand_do_upgrade "$1"
}
```
where it expects the folder inside sysupgrade.bin containing the kernel to be named "sysupgrade-routerboard/". If I don't add the `BOARD_NAME` parameter, the folder is named after the board (i.e., mikrotik,routerboard-922uags-5hpacd). I can pass the board name as a parameter, but then, downgrading to ar71xx does not work. Therefore, there are two options:

1. To add the `BOARD_NAME` as it was in ar71xx
2. To modify the `platform_do_upgrade_mikrotik_rb()`function as I did here: https://github.com/rogerpueyo/openwrt/commit/607663f9ba5db1e1d0ceeda71caf75831f0d1b0b#diff-aeec2f44188773e103204c76d93ece4eR22-R28

Probably the first option is the quick workaround, while the second is the way to go, since in ath79 each MikroTik board will have a different firmware file?

Reply:
Hmm, I'd say we do not support downgrade.
BOARD_NAME is not used a single time in ath79 or ramips, except for this workaround in 19.07: https://github.com/openwrt/openwrt/commit/a3ffeb413bca3da273aac924f4aa0f326a66a620
As I understood it, it is a legacy concept.

So, I personally would choose neither 1. nor 2.

However, maybe other people have different opinions here.

Reply:
I just realised that not adding the `BOARD_NAME` setting means that upgrading from ar71xx to ath79 is neither possible:
```
root@(none):/# sysupgrade -F -v -n /tmp/openwrt-ath79-nand-mikrotik_routerboard-
922uags-5hpacd-squashfs-sysupgrade.bin 
Invalid sysupgrade file.
Image check 'platform_check_image' failed but --force given - will update anyway!
Commencing upgrade. Closing all shell sessions.
Watchdog handover: fd=3
- watchdog -
root@OpenWrt:/# killall: telnetd: no process killed
Killed
Sending TERM to remaining processes ... sh sleep 
Sending KILL to remaining processes ... 
Switching to ramdisk...
Performing system upgrade...
Unlocking kernel ...
Erasing kernel ...
Writing data to block 0 at offset 0x0
tar: sysupgrade-routerboard/kernel: not found in archive
removing ubiblock0_1
[  345.641079] block ubiblock0_1: released
Volume ID 0, size 16 LEBs (2031616 bytes, 1.9 MiB), LEB size 126976 bytes (124.0 KiB), dynamic, name "none", alignment 1
Volume ID 1, size 26 LEBs (3301376 bytes, 3.1 MiB), LEB size 126976 bytes (124.0 KiB), dynamic, name "rootfs", alignment 1
Set volume size to 117579776
Volume ID 2, size 926 LEBs (117579776 bytes, 112.1 MiB), LEB size 126976 bytes (124.0 KiB), dynamic, name "rootfs_data", alignment 1
sysupgrade successful
umount: can't unmount /tmp: Resource busy
umount: can't unmount /dev: Resource busy
[  348.115242] reboot: Restarting system
```

The important line above is `tar: sysupgrade-routerboard/kernel: not found in archive`, since the kernel route in the TAR file is now `sysupgrade-mikrotik_routerboard-922uags-5hpacd/kernel`.

How should I address this? How important is having sysupgrade from ar71xx?

Reply:
sysupgrade from ar71xx would be good. Still, I wonder that this is a problem only for this device, which IIRC would be the first in entire ath79 to use BOARD_NAME again.
Note that upgrading from 18.06 might behave differently here than upgrading from 19.07 ar71xx.

However, I'm not really familiar with the special measures for updating nand devices, so I can not discuss this in detail.

Reply:
I think for the platform_do_upgrade_mikrotik_rb() issue one can pass in the $board variable as an extra argument, and then replace the sysupgrade-routerboard/kernel text with "sysupgrade-${2/,/_}/kernel", and yes it should be quoted.

The ${2/,/_} shell syntax means "replace a ',' with a '_' in the 2nd variable.

Reply:
Just found the following:
https://github.com/openwrt/openwrt/blob/master/package/base-files/files/lib/upgrade/nand.sh#L249

```
	local board_dir=$(tar tf $tar_file | grep -m 1 '^sysupgrade-.*/$')
	board_dir=${board_dir%/}

	local kernel_length=`(tar xf $tar_file ${board_dir}/kernel -O | wc -c) 2> /dev/null`
	local rootfs_length=`(tar xf $tar_file ${board_dir}/root -O | wc -c) 2> /dev/null`
```

So, to me it looks like we could just use
```
	local board_dir=$(tar tf "$1" | grep -m 1 '^sysupgrade-.*/$')
	board_dir=${board_dir%/}
	tar xf "$1" ${board_dir}/kernel -O | nandwrite -o "$fw_mtd" -
```

and that would even keep downgrade available?